src/repeated_game.jl

```diff
@@ -0,0 +1,339 @@
+    # Objective depends on which player we are minimizing
+    c = zeros(2)
+    c[i] = 1.0
+
+    # Lower and upper bounds for w
+    lb = [-Inf, -Inf]
+    ub = [Inf, Inf]
+
+    lpout = linprog(c, H, '<', C, lb, ub, ClpSolver())
+    if lpout.status == :Optimal
+        out = lpout.sol[i]
+    else
+        out = minimum(rpd.sg.players[i].payoff_array)
+    end
+
+    return out
+end
+
+worst_value_1(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}) =
+    worst_value_i(rpd, H, C, 1)
+worst_value_2(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}) =
+    worst_value_i(rpd, H, C, 2)
+worst_values(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}) =
+    (worst_value_1(rpd, H, C), worst_value_2(rpd, H, C))
+
+#
+# Outer Hyper Plane Approximation
+#
+"""
+Approximates the set of equilibrium value set for a repeated game with the
+outer hyperplane approximation described by Judd, Yeltekin, Conklin 2002
+"""
+function outerapproximation(rpd::RepeatedGame; nH=32, tol=1e-8, maxiter=500, nskipprint=1)
+    # Long unpacking of stuff
+    sg, δ = unpack(rpd)
+    p1, p2 = sg.players
+    po_1, po_2 = p1.payoff_array, p2.payoff_array
+    p1_minpayoff, p1_maxpayoff = extrema(po_1)
+    p2_minpayoff, p2_maxpayoff = extrema(po_2)
+
+    # Get number of actions for each player
+    nA1, nA2 = num_actions(p1), num_actions(p2)
+    nAS = nA1 * nA2
+
+    # Create action space
+    AS = QuantEcon.gridmake(1:nA1, 1:nA2)
+
+    # Create the unit circle, points, and hyperplane levels
+    C, H, Z = create_unitcircle_points(rpd, nH)
+    Cnew = copy(C)
+
+    # Create matrices for linear programming
+    c, A, b = initialize_LP_matrices(rpd, H)
+
+    # bounds on w are [-Inf, Inf] while bounds on slack are [0, Inf]
+    lb = vcat([-Inf, -Inf], zeros(nH+2))
+    ub = fill(Inf, nH+4)
+
+    # Set iterative parameters and iterate until converged
+    iter, dist = 0, 10.0
+    while (iter < maxiter) && (dist > tol)
+        # Compute the current worst values for each agent
+        _w1 = worst_value_1(rpd, H, C)
+        _w2 = worst_value_2(rpd, H, C)
+
+        # Update all set constraints
+        copy!(b, 1, C)
+
+        # Iterate over all subgradients
+        for ih=1:nH
+            #
+            # Subgradient specific instructions
+            #
+            h1, h2 = H[ih, :]
+
+            # Put the right objective into c (negative because want maximize)
+            c[1] = -h1
+            c[2] = -h2
+
+            # Allocate space to store all solutions
+            Cia = Array(Float64, nAS)
+            Wia = Array(Float64, 2, nAS)
+            for ia=1:nAS
+                #
+                # Action specific instruction
+                #
+                a1, a2 = AS[ia, :]
+
+                # Update incentive constraints
+                b[nH+1] = flow_u_1(rpd, a1, a2) - best_dev_payoff_1(rpd, a2) - δ*_w1
+                b[nH+2] = flow_u_2(rpd, a1, a2) - best_dev_payoff_2(rpd, a1) - δ*_w2
```

Comment:
I am afraid the definition of `flow_u_*` is confusing. I would define it as the value without `(1-δ)` (flow value!),
and then multiply it by `(1-δ)` here. (Then `flow_u_*` would depend only on the underlying `NormalFormGame` instead of `rpd`.) Same for `best_dev_payoff_*`.